### PR TITLE
feature edition: geom validation for lines and polygons

### PR DESCRIPTION
**Pull request by matallo**
re: https://github.com/CartoDB/cartodb/issues/11052

- added geometry validation for lines and polygons
- added the corresponding tests

this uses a cartodb.js where some geom helpers are exposed, maybe some of the model methods can be extracted to those helpers

screen shot 2017-06-28 at 17 20 09
screen shot 2017-06-28 at 17 20 25

**acceptance**

draw a geometry (points, lines, polygons) inside the bounds:
it should pass

draw geometries (points, lines, polygons) outside bounds (pan the map for this purpose):
an error tooltip should appear in the field

### CODE REVIEW

**Comment by xavijam** [comment]
Tests are not passing @matallo

**Comment by matallo** [comment]
retest this, please

File: lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js [hunk 1]
```diff
@@ -10,12 +19,86 @@ module.exports = Backbone.Model.extend({
...
+
+    var geoJSON = JSON.parse(value);
+
+    if (this._featureModel.isLine()) {
+      coordinates = cdb.helpers.GeoJSONHelper.getPolylineLatLngsFromGeoJSONCoords(geoJSON);
+    } else {
+      coordinates = cdb.helpers.GeoJSONHelper.getPolygonLatLngsFromGeoJSONCoords(geoJSON);
+    }
+
+    for (var i = 0, l = coordinates.length; i < l; i++) {
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js]: I hardly see a `for` loop these days, here is an alternative more functional way (just for fun):

```
var errors = coordinates.reduce(function (memo, coordinate) {
  var latitude = this._getLatitude(position);
  var longitude = this._getLongitude(position);

  var latError = (latitude > this._COORDINATES_OPTIONS.max_lat || latitude < this._COORDINATES_OPTIONS.min_lat);
 
  if (latError) {
    memo.push(_t('editor.edit-feature.out-of-bounds-lat'));
  }

  var lngError = (longitude > this._COORDINATES_OPTIONS.max_lng || longitude < this._COORDINATES_OPTIONS.min_lng);

  if (lngError) {
    memo.push(_t('editor.edit-feature.out-of-bounds-lat'));
  }
}, []).filter(function(item, pos, self) {
    return self.indexOf(item) === pos;
});

var validation = null;

if (errors.length > 0) {
  validation = {
    message: errors.join(' ');
  }
}

return validation;
```

File: lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js [hunk 2]
```diff
@@ -10,12 +19,86 @@ module.exports = Backbone.Model.extend({
...
+      var latitude = this._getLatitude(position);
+      var longitude = this._getLongitude(position);
+
+      if (!latError) {
+        latError = (latitude > this._COORDINATES_OPTIONS.max_lat || latitude < this._COORDINATES_OPTIONS.min_lat);
+      }
+
+      if (!lngError) {
+        lngError = (longitude > this._COORDINATES_OPTIONS.max_lng || longitude < this._COORDINATES_OPTIONS.min_lng);
+      }
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js]: I'd add a `break` here if `(lngError || latError)`.

File: lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js (hunk 1 above)
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/edit-feature-content-views/edit-feature-geometry-form-model.js]: love that reduce 😍

**Comment by matallo** [comment]
@nobuti I just went all functional `_.chain().reduce().uniq().value()`

I was being lazy last Friday so you might have received some failed builds errors, as I pushed directly without testing, tests should be 💚 now.

Also I made some change where `_formatCoordinate` in `_getLatitude` and `_getLongitude` methods were returning `undefined` for value "0".

**Comment by nobuti** [comment]
Love it!

**Comment by nobuti** [comment]
I found an edge case, but I'm afraid it's hard to fix, at least I don't know how:

feature

After talking to @matallo and a thorough investigation, we accord that it's difficult to fix it due geometry doesn't report useful information once the feature is outside the "canvas".

Otherwise, it works wonderfully ✨ . For me, it's a green light, thoughts @xavijam?

**Comment by xavijam** [comment]
I feel it should relay on CartoDB.js "edition" feature. Could we open a ticket there and go on with this? 🙏

**Comment by matallo** [comment]
indeed, that node belongs to cartodb.js, not really much to do besides the work already done

**Comment by nobuti** [comment]
Done here: https://github.com/CartoDB/cartodb.js/issues/1692. Now 🚀 this.

**Comment by nobuti** [comment]
In production.

**Pull request merged**